janusgraph-core/src/main/java/org/janusgraph/graphdb/tinkerpop/optimize/strategy/JanusGraphMultiQueryStrategy.java

```diff
@@ -0,0 +1,75 @@
+// Copyright 2021 JanusGraph Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package org.janusgraph.graphdb.tinkerpop.optimize.strategy;
+
+import org.apache.tinkerpop.gremlin.process.traversal.Traversal.Admin;
+import org.apache.tinkerpop.gremlin.process.traversal.TraversalStrategy;
+import org.apache.tinkerpop.gremlin.process.traversal.strategy.AbstractTraversalStrategy;
+import org.apache.tinkerpop.gremlin.process.traversal.strategy.decoration.SubgraphStrategy;
+import org.apache.tinkerpop.gremlin.process.traversal.util.TraversalHelper;
+import org.apache.tinkerpop.gremlin.structure.Graph;
+import org.janusgraph.graphdb.database.StandardJanusGraph;
+import org.janusgraph.graphdb.tinkerpop.optimize.JanusGraphTraversalUtil;
+import org.janusgraph.graphdb.tinkerpop.optimize.step.JanusGraphMultiQueryStep;
+import org.janusgraph.graphdb.tinkerpop.optimize.step.MultiQueriable;
+import org.janusgraph.graphdb.transaction.StandardJanusGraphTx;
+
+import java.util.Collections;
+import java.util.Set;
+
+/**
+ * @author Marko A. Rodriguez (https://markorodriguez.com)
+ * @author Matthias Broecheler (http://matthiasb.com)
+ */
+public class JanusGraphMultiQueryStrategy extends AbstractTraversalStrategy<TraversalStrategy.ProviderOptimizationStrategy> implements TraversalStrategy.ProviderOptimizationStrategy {
+
+    private static final Set<Class<? extends ProviderOptimizationStrategy>> PRIORS = Collections.singleton(JanusGraphLocalQueryOptimizerStrategy.class);
+    private static final JanusGraphMultiQueryStrategy INSTANCE = new JanusGraphMultiQueryStrategy();
+
+    private JanusGraphMultiQueryStrategy() {
+    }
+
+    @Override
+    public void apply(final Admin<?, ?> traversal) {
+        if (!traversal.getGraph().isPresent()
+            || TraversalHelper.onGraphComputer(traversal)
+            || TraversalHelper.getRootTraversal(traversal) != traversal
```

Comment:
For the record: I'm currently trying to implement a prototype of the second solution where unnecessary steps could remain in the traversal.
To mitigate that problem, my plan is to change the relation between `JanusGraphMultiQueryStep` and `MultiQueriable` from pull based to push based. Currently, each `JanusGraphMultiQueryStep` aggregates traversers in a cache, which gets cleared from time to time by a `MultiQueriable` client. This is what I refer to as "pull based" and it causes problems because unnecessary `JanusGraphMultiQuerySteps` would never get cleared and therefore cause OOM.
In contrast, the push based architecture would mean that the cache resides within the `MultiQueriable`. Each `JanusGraphMultiQueryStep` could have a list of known clients and push traversers into their caches. This way, an orphan `JanusGraphMultiQueryStep` would simply not have any client and therefore do no work and cause no harm.

Reply:
Likely to be resolved with #2619